library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java

```diff
@@ -40,20 +42,62 @@ public BatchFileBuilder saveTo(String path) {
 
     @Override
     public BatchFileBuilder saveTo(String path, String fileName) {
-        this.path = Optional.fromNullable(path).filterNot("/"::equals);
+        this.path = Optional.fromNullable(path);
         this.fileName = Optional.fromNullable(fileName);
         return this;
     }
 
     @Override
     public BatchBuilder apply() {
-        String absolutePath = storageRoot.path() + File.separator
-                + downloadBatchId.rawId() + File.separator
-                + path.map((it) -> it + File.separator).or("")
-                + fileName.getOrElse(() -> FileNameExtractor.extractFrom(networkAddress));
+        String absolutePath = buildPath(
+                storageRoot.path(),
+                downloadBatchId.rawId(),
+                path.or(""),
+                fileName.getOrElse(() -> FileNameExtractor.extractFrom(networkAddress))
+        );
 
         parentBuilder.withFile(new BatchFile(networkAddress, downloadFileId, absolutePath));
         return parentBuilder;
     }
 
+    private String buildPath(String... paths) {
+        StringBuilder stringBuilder = new StringBuilder();
+
+        String storagePath = paths[0];
+        if (storagePath.startsWith(File.separator)) {
+            stringBuilder.append(File.separator);
+        }
+
+        String joinedPaths = sanitise(joinWithFileSeparator(paths));
+        stringBuilder.append(joinedPaths);
+
+        return stringBuilder.toString();
+    }
+
+    private String sanitise(String path) {
+        String[] pathSegments = path.split("/");
+        CharSequence[] filteredPathSegments = filterEmptySegmentsOut(pathSegments);
+        return joinWithFileSeparator(filteredPathSegments);
+    }
+
+    private CharSequence[] filterEmptySegmentsOut(String[] pathSegments) {
+        List<CharSequence> filteredPathSegments = new ArrayList<>();
+        for (String pathSegment : pathSegments) {
+            if (!pathSegment.isEmpty()) {
+                filteredPathSegments.add(pathSegment);
+            }
+        }
+        return filteredPathSegments.toArray(new CharSequence[0]);
+    }
+
+    private String joinWithFileSeparator(CharSequence[] elements) {
```

Comment:
I opted to re-implement the `join` method since the `TextUtils` one can't be used in tests and `String.join` is definitely not available for our min API (16).